lib/daru/io/io.rb

```diff
@@ -201,25 +199,29 @@ def from_csv_prepare_opts opts
 
       def from_csv_hash_with_headers(path, opts)
         opts[:header_converters] ||= :symbol
-
         ::CSV
-          .read(path, 'rb',opts)
+          .parse(new_open(path), opts)
           .tap { |c| yield c if block_given? }
           .by_col.map { |col_name, values| [col_name, values] }.to_h
       end
 
       def from_csv_hash(path, opts)
         csv_as_arrays =
           ::CSV
-          .open(path, 'rb', opts)
+          .parse(new_open(path), opts)
           .tap { |c| yield c if block_given? }
           .to_a
-
         headers       = ArrayHelper.recode_repeated(csv_as_arrays.shift)
         csv_as_arrays = csv_as_arrays.transpose
-
         headers.each_with_index.map { |h, i| [h, csv_as_arrays[i]] }.to_h
       end
+
+      def new_open(url)
+        open(url)
+      rescue URI::InvalidURIError
+        host = url.match(".+\:\/\/([^\/]+)")[1]
+        Net::HTTP.get host, url.partition(host)[2] || '/'
```

Comment:
Have a look at this issue - [Open-URI can't open URLs with underscores](https://bugs.ruby-lang.org/issues/8241). Though the issue has been closed, the scheme has been updated only from Ruby 2.2 onwards - but our Travis CI build takes earlier versions too into consideration.

Rather, this workaround `new_open` can open URLs with underscores too. Though this might be a discrete case, I wouldn't want to assume that NO user will ever use excel files from URLs that have underscores.

I am thinking of renaming the `new_open` function to `open_with_underscore` - please feel free to share your choice for the function name.